README.md
# pykota
Git Repository to Pykota

This repository was created to update the pykota project and adapt it to Python version 3.10 or higher.

### Features
 - Updated the version to work with python version 3.10 or higher;

### Considerations
 - This version was tested only on Postgres Database;
 - We adapt in order to meet personal demands;
 - Only the pkipplib, pkpgcounter and pykota projects have been migrated to version 3.10 or higher;
 - The tea4cups project have not migrated at the moment and we don't know if we will update. The pykotaicon project will be updated soon.





### Simple tutorial - Instalation Pykota with Ubuntu Server 23.04 and PostgreSQL

#### With fresh install Ubuntu Server 23.04 and upgrade all packages ... and get power of root.. let's go.


Clone repository:

`# cd /opt`

`# git clone https://github.com/heana-hosp/pykota.git`


*Get the power of root and proceed to install - every command with permission*

Install CUPS Service:

`# apt install cups `

Easy way to get exception on apparmor for CUPS:

`apt install apparmor-utils`

`aa-complain cupsd`

`/etc/init.d/apparmor restart`

Check version of python - must be equal to or higher than 3.10:

`# python3 --version`

Install PostgreSQL:

`apt install postgresql postgresql-contrib`

Install Python libs needed to this version:

`# apt install python3-pil python3-pypdf2 libpq-dev python3.11-dev python3-psycopg2 python3-reportlab`


#### Build and Install Pykota libs, tools, configuration and database.

1 - Install pkipplib:

`# cd /opt/pykota/pkipplib/`

`# python3 setup.py build`

`# python3 setup.py install`

2 - Install pkpgcounter:

`# cd /opt/pykota/pkpgcounter/`

`# python3 setup.py build`

`# python3 setup.py install`

3 - Install pykota:

Check dependencies before:

`# cd /opt/pykota/pykota/`

`# python3 checkdeps.py`

*Check the output if some libs were detected - in this case, the library for postgres and pkpgcounter(in previous step) was correctly installed.*

```batch
...
Checking for Python-Psycopg availability : OK
...
...
Checking for Python-pkpgcounter availability : OK
...
```

Build and Install:

`# python3 setup.py build`

`# python3 setup.py install`

Generate database:

`# su - postgres -c "psql -f  /opt/pykota/pykota/initscripts/postgresql/pykota-postgresql.sql template1"`

*By default, the database passwords are already defined for the pykotauser and pykotaadmin users(after generate script). If you need to change the password, change it at the database and don't forget to update passwords on files pykota.conf and pykotadmin.conf on directory /etc/pykota (instructions ahead).*


Copy backend from pykota to cups libs and give permissions:

`# cp /opt/pykota/pykota/bin/cupspykota /usr/lib/cups/backend/`

`# chmod -R 755 /usr/lib/cups/backend/cupspykota`

`# chmod +x /usr/lib/cups/backend/cupspykota`

Restart Cups:

`# systemctl restart cups`

Create user pykota on system:

`adduser --system --group --home /etc/pykota --gecos PyKota pykota`

Copy default configuration files and set permission:

`cp /opt/pykota/pykota/conf/pykota.conf.sample /etc/pykota/pykota.conf`

`cp /opt/pykota/pykota/conf/pykotadmin.conf.sample /etc/pykota/pykotadmin.conf`

`chmod 755 /etc/pykota`

`chmod 644 /etc/pykota/pykota.conf`

`chmod 644 /etc/pykota/pykotadmin.conf`

`chown pykota:pykota /etc/pykota/pykota.conf /etc/pykota/pykotadmin.conf`


Adjust the correct path to pkpgcounter and pknotify in the /etc/pykota/pykota.conf file

Check the path:
```bash
# which pkpgcounter
/usr/local/bin/pkpgcounter
# which pknotify
/usr/local/bin/pknotify
```

Change lines or comment - pknotify will return to client a prompt message(pykotaicon on client):

```bash
...
accounter : software(/usr/local/bin/pkpgcounter)
...
preaccounter: software(/usr/local/bin/pkpgcounter)
...
askconfirmation : /usr/local/bin/pknotify --destination $PYKOTAJOBORIGINATINGHOSTNAME:7654 --timeout 7 --confirm "Hello $PYKOTAUSERNAME.\nThe job  $PYKOTAJOBID will be send to $PYKOTAPRINTERNAME and cost $PYKOTAPRECOMPUTEDJOBSIZE page(s)\nYour balance is $PYKOTABALANCE\n\nDo you really want to print?"

```

#### Configure CUPS and add printer.

Reference Guide: https://ubuntu.com/server/docs/service-cups

Change JobPrivateAccess to **all** on `<Policy default>`

Change JobPrivateValues to **none** on `<Policy default>`

On `<Location />` and `<Location /admin>` add **Allow IP_STATION_WILL_CONFIGURE_PRINTERS**

Example: If I want to access the server remotely from IP 100.100.100.133 - follow the configuration section of the file:

```bash
...
...
<Location />
  Order allow,deny
  Allow 100.100.100.133
</Location>

# Restrict access to the admin pages...
<Location /admin>
  Order allow,deny
  Allow 100.100.100.133
</Location>

# Restrict access to configuration files...
<Location /admin/conf>
  AuthType Default
  Require user @SYSTEM
  Order allow,deny
</Location>

# Restrict access to log files...
<Location /admin/log>
  AuthType Default
  Require user @SYSTEM
  Order allow,deny
</Location>

# Set the default printer/job policies...
<Policy default>
  # Job/subscription privacy...
  JobPrivateAccess all
  JobPrivateValues none
  SubscriptionPrivateAccess default
  SubscriptionPrivateValues default
...
... 
```

Add lpadmin group

`# usermod -a -G  lpadmin $USER`

Restart CUPS

`systemctl restart cups`

Add printer on CUPS:

In the browser(emember the IP assigned in the cups configuration file) access **IP_SERVER:631**

**Administration -> Add Printer -> Login with user from linux(root) and password -> Option: Other Network Printers -> IPP -> Continue;**

Add backend cupspykota on connection: change ipp to **cupspykota:ipp://IP_PRINTER** -> continue to next step;

Give a Name, Description, Localization and mark Share this printer and continue. **Important:** The name is sensitive case, and will be used the same text to define printer on pykota.

Configure the driver for your print. Recomended the postscripts/cupsfilter drivers. You can find optimized driver ppd on internet for specified printer. Finally complete the installation.

#### Add user and printer on pykota

Add printer(remember, case sensitive: same name of printer installed on cups). Ex. name is TEST:

`# pkprinters --add TEST`

Add user and set balance. In this case, the user is the same user who is logged in to the Windows station. To find out who the user is. just run (echo %USERNAME%) in cmd. In this example we will use **testuser**:

`# pkusers --add usertest`

Defining the balance type for the user and defining a balance. Pykota allows you to assign the balance and quota type. Next step, give 50 balance.

`# pkusers --limitby balance usertest`

`# pkusers --balance +50.0 --comment "add more 50!" usertest`

Link user: testuser to printer: TEST

`# edpykota --add usertest`

`# edpykota --printer TEST -S 50 -H 50 usertest`


#### Add printer in Windows

win + r (Execute) -> control printers

Add printer -> The printer is not in the list ->  Select Printer Shared by Name:

format: `http://IP_SERVER:631/printers/NAME_PRINTER_CREATED_CUPS`

Ex. http://IP_SERVER:631/printers/TEST

After proceeding, you will be directed to the driver installation wizard and complete instalation.

#### Execute pykoticon on host Windows

Download example client windows from oficial site:

http://www.pykota.com/software/pykoticon/download/tarballs/pykoticon-1.02.zip

Extract to any directory

Edit the file pykoticon.vbs

Change the line `retcode = Wshshell.run ("%comspec% /C pykoticon.exe --port 7654 nordine.ateur", 0, FALSE)`

Assign the IP of the server you just installed: `retcode = Wshshell.run ("%comspec% /C pykoticon.exe --port 7654 IP_SERVER", 0, FALSE)`

Save and run the application pykoticon.exe. The icon will appear in the tray icon. Now just send a print for testing.



































tea4cups/NEWS
# $Id: NEWS 3576 2018-08-23 00:29:57Z jerome $

Tea4CUPS : Tee for CUPS

(c) 2005-2018 Jerome Alet <alet@librelogiciel.com>
(c) 2005 Peter Stuge <stuge-tea4cups@cdy.org>
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

=============================================================================

Tea4CUPS News :

  * 3.14alpha (2018-08-23) :

    - Fixes a problem where the job billing attribute sometimes is an integer.
    
  * 3.13alpha (2006-11-07) :

    - Backported locking mechanism from PyKota.

  * 3.12 (2006-08-11) :

    - Serializes accesses to the same device from different queues
      or print servers through file locking facilities (works over
      NFS).

    - Improved support for CUPS 1.2.x and higher.

    - Allows the administrator to cleanly stop tea4cups with SIGINT.

    - Added a configurable retry directive to the configuration file.

    - Fixed several minor bugs.

    - Improved the installation instructions and the sample configuration
      file.

  * 3.11 :

    - Fixed an incompatibility with Python 2.1, thanks to Frank
      Koormann.

  * 3.10 :

    - Added filters.

    - Added the "onfail" directive to tea4cups.conf

  * 3.02 :

    - Fixed some problems thanks to pychecker.

  * 3.01 :

    - Fixed an IPP parsing bug in some situations.

  * 3.00 :

    - Tees don't exist anymore. Only prehooks and posthooks remain.

    - Prehooks can now send datas to Posthooks through pipes.

    - Major rewrite of the subprocess management code, thanks to Peter Stuge.

  * 2.12alpha :

    - Greatly improved the IPP parser.

    - When printing test pages from CUPS' web interface, the empty username
      is now overwritten with the username CUPS is running as.

  * 2.11 :

    - Now works with old releases of Python (>= 2.1)

    - More powerful error handling at startup time.

    - Ignores errors in posthooks because the job was probably
      successfully printed so no need to stop the print queue.
      We log a warning instead in this case.

  * 2.10 :

    - Makes the IPP message filename and the Job's billing code available.

  * 2.00 :

    - Introduces the notions of pre and post hooks.

    - Prehooks can decide to entirely cancel the print job if needed.

    - Original CUPS backend's exit code is available in posthooks.

    - prehooks, posthooks and tees are now launched in alphabetical order.

  * 1.00 :

    - First Official release.



pykota/checkdeps.py
#! /usr/bin/env python3
# -*- coding: ISO-8859-15 -*-
#
# PyKota
#
# PyKota : Print Quotas for CUPS and LPRng
#
# (c) 2003, 2004, 2005, 2006, 2007 Jerome Alet <alet@librelogiciel.com>
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
#
# $Id: checkdeps.py 3133 2007-01-17 22:19:42Z jerome $
#
#

import sys
import os


def checkModule(module):
    """Checks if a Python module is available or not."""
    try:
        exec(f"import {module}")
    except ImportError:
        return 0
    else:
        return 1


def checkCommand(command):
    """Checks if a command is available or not."""
    input = os.popen(f"type {command} 2>/dev/null")
    result = input.read().strip()
    input.close()
    return result


def checkWithPrompt(prompt, module=None, command=None, helper=None):
    """Tells the user what will be checked, and asks him what to do if something is absent."""
    sys.stdout.write(f"Checking for {prompt} availability : ")
    sys.stdout.flush()
    if command is not None:
        result = checkCommand(command)
    elif module is not None:
        result = checkModule(module)
    if result:
        sys.stdout.write("OK\n")
    else:
        sys.stdout.write("NO.\n")
        sys.stderr.write(f"ERROR : {prompt} not available !\n")
        sys.stdout.write("%s\n" % helper)


if __name__ == "__main__":
    print("Checking PyKota dependencies...")

    # checks if Python version is correct, we need >= 2.2
    if not (sys.version > "3.1"):
        sys.stderr.write(
            "PyKota needs at least Python v3.1 !\nYour version seems to be older than that, please update.\nAborted !\n")
        sys.exit(-1)

    # checks if some needed Python modules are there or not.
    modulestocheck = [("Python-Psycopg", "psycopg2",
                       "PygreSQL is mandatory if you want to use PostgreSQL as the quota database backend.\nfor debian or ubuntu apt install libpq-dev python3.10-dev\nSee http://www.pygresql.org"),
                      ("Python-SQLite", "pysqlite2",
                       "Python-SQLite is mandatory if you want to use SQLite as the quota database backend.\nSee http://www.pysqlite.org"),
                      ("MySQL-Python", "MySQLdb",
                       "MySQL-Python is mandatory if you want to use MySQL as the quota database backend.\nSee http://sourceforge.net/projects/mysql-python"),
                      ("Python-datetime", "datetime",
                       "eGenix' Datetime is mandatory for PyKota to work.\nSee http://www.egenix.com"),
                      ("Python-LDAP", "ldap",
                       "Python-LDAP is mandatory if you plan to use an LDAP\ndirectory as the quota database backend.\nSee http://python-ldap.sf.net"),
                      ("Python-OSD", "pyosd",
                       "Python-OSD is recommended if you plan to use the X Window On Screen Display\nprint quota reminder named pykosd. See http://repose.cx/pyosd/"),
                      ("Python-SNMP", "pysnmp",
                       "Python-SNMP is recommended if you plan to use hardware\naccounting with printers which support SNMP.\nSee http://pysnmp.sf.net"),
                      ("Python-JAXML", "jaxml",
                       "Python-JAXML is recommended if you plan to dump datas in the XML format.\nSee http://www.librelogiciel.com/software/"),
                      ("Python3-ReportLab", "reportlab.pdfgen.canvas",
                       "Python-ReportLab is required if you plan to have PyKota generate banners.\nSee http://www.reportlab.org/"),
                      ("Python-Imaging", "PIL.Image",
                       "Python-Imaging is required if you plan to have PyKota generate banners.\nSee http://www.pythonware.com/downloads/"),
                      ("Python-Psyco", "psyco",
                       "Python-Psyco speeds up parsing of print files, you should use it.\nSee http://psyco.sourceforge.net/"),
                      ("Python-pkpgcounter", "pkpgpdls",
                       "Python-pkpgcounter is mandatory.\nGrab it from http://www.pykota.com/software/pkpgcounter/"),
                      ("Python-PAM", "PAM",
                       "Python-PAM is recommended if you plan to use pknotify+PyKotIcon.\nGrab it from http://www.pangalactic.org/PyPAM/"),
                      ("Python-pkipplib", "pkipplib",
                       "Python-pkipplib is now mandatory.\nGrab it from http://www.pykota.com/software/pkipplib/"),
                      ("Python-chardet", "chardet",
                       "Python-chardet is recommended.\nfor debian or ubuntu run pip3.10 install chardet\nGrab it from http://chardet.feedparser.org/"),
                      ]
    commandstocheck = [("GhostScript", "gs",
                        "Depending on your configuration, GhostScript may be needed in different parts of PyKota."),
                       ("SNMP Tools", "snmpget", "SNMP Tools are needed if you want to use SNMP enabled printers."),
                       ("Netatalk", "pap", "Netatalk is needed if you want to use AppleTalk enabled printers.")
                       ]
    for (name, module, helper) in modulestocheck:
        checkWithPrompt(name, module=module, helper=helper)

    # checks if some software are there or not.
    for (name, command, helper) in commandstocheck:
        checkWithPrompt(name, command=command, helper=helper)
